### PR TITLE
Add poll_timeout attribute to checkin endpoint

### LINKED ISSUE #2337: Increase long poll default to ~30 minutes
We've been doing scale testing over the past few months using a ~30 minute long poll duration (rather than current default of 5m) and we are seeing much better results for very large clusters.

We're now ready to make this the default setting for Fleet Server. A corresponding change to Agent will be made, I'll link the issue soon.

### Tasks

The timeout at the proxy will be slightly over 30 minutes at 30m 20s. We should configure Fleet Server to timeout shortly before that, so we'll use a 28m long poll duration with a 29m response write timeout.

- [ ] Update default for `server.timeouts.checkin_long_poll` to 28m here: https://github.com/elastic/fleet-server/blob/c8527d3d9f55c44d4f4764a017c93731f595f342/internal/pkg/config/timeouts.go#L57
- [ ] Update default for `server.timeouts.write` to 29m here: https://github.com/elastic/fleet-server/blob/c8527d3d9f55c44d4f4764a017c93731f595f342/internal/pkg/config/timeouts.go#L47-L51

### PR DESCRIPTION
**Pull request by michel-laterman**
## What is the problem this PR solves?

Allow clients to request longer polls for checkin requests.
This results in better performance in larger deployments

The approach is backwards compatible as no defaults are changed.
Changing the default times to larger values may result in older agents requests timing out.

## How does this PR solve the problem?

Add a `poll_duration` string attribute to checkin request bodies that inform's fleet-server of the client's poll duration. And a `checkin_max_poll` config option to enforce an upper limit to requested times.

If present the `poll_duration` string is parsed into a `time.Duration`, if the result is a non-zero value fleet-server will use the provided `poll_duration` to set it's poll-time and write deadline for the request

Fleet-server will set it's poll time to `min(max(1m, poll_duration-2m), checkin_max_poll)` before doing the normal setup/jitter offsets.
And the request's write deadline will be set to `max(2m, poll time + 1m)`.
This leads to the min values a poll duration of 1m and the write timeout to 2m.
If `checkin_max_poll` is set to less then 1m it is effectively ignored and a 1m value is used.

The min accepted `poll_time` from the client is 3m, and the max is `checkin_max_poll`-2m

If `poll_duration` is not set (or parses to a zero value) the configured checkin values will be used instead (defaults to a 5m poll and 10m write deadline)

If `poll_duration` cannot be parsed an error is returned.

## Design Checklist

- [x] I have ensured my design is stateless and will work when multiple fleet-server instances are behind a load balancer.
- ~~I have or intend to scale test my changes, ensuring it will work reliably with 100K+ agents connected.~~ will require client changes to test at scale
- ~~I have included fail safe mechanisms to limit the load on fleet-server: rate limiting, circuit breakers, caching, load shedding, etc.~~

## Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Closes #2337

### CODE REVIEW

File: internal/pkg/api/handleCheckin.go [hunk 1]
```diff
@@ -108,26 +108,51 @@ func (ct *CheckinT) handleCheckin(zlog zerolog.Logger, w http.ResponseWriter, r
...
+		}
+	}
+
+	pollDuration := ct.cfg.Timeouts.CheckinLongPoll
+	// set the pollDuration if pDur parsed from poll_timeout was a non-zero value
+	// sets timeout to max(1m, pDur-2m)
+	// sets the response write timeout to max(2m, pDur-1m)
+	if pDur != time.Duration(0) {
+		pollDuration = pDur - (2 * time.Minute)
+		if pollDuration < time.Minute {
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/api/handleCheckin.go:142]: I agree, please add a configuration setting with an upper limit.

**Review by blakerouse — APPROVED** [review]
Overall this looks really good. I agree with @jsoriano on adding an upper limit.

File: model/openapi.yml [hunk 2]
```diff
@@ -255,6 +255,13 @@ components:
           type: string
           format: application/json
           x-go-type: json.RawMessage
+        poll_timeout:
+          description: |
+            An optional timeout value that informs fleet-server of when a client will time out on it's checkin request.
+            If not specified fleet-server will use the timeout values specified in the config (defaults to 5m polling and a 10m write timeout).
+            If specified fleet-server will set its poll timeout to `max(1m, poll_timeout-2m)` and its write timeout to `max(2m, poll_timout-1m)`.
```
[the remark this answers is not in the document]
Comment (michel-laterman) [reply model/openapi.yml:262]: We want to have the write timeout after poll timeout so that the fleet-server still has time to write the response for a long running poll.

File: internal/pkg/config/timeouts.go [hunk 3]
```diff
@@ -58,4 +59,9 @@ func (c *ServerTimeouts) InitDefaults() {
 
 	// Jitter subtracted from c.CheckinLongPoll.  Disabled if zero.
 	c.CheckinJitter = 30 * time.Second
+
+	// MaxPoll is the maximum allowed value for a long poll when the client specified poll_timeout value is used.
+	// The long poll value is poll_timeout-2m, and the request's write timeout is set to poll_timeout-1m
+	// CheckinMaxPoll values of less then 1m are effectively ignored and a 1m limit is used.
+	c.CheckinMaxPoll = 2 * time.Hour
```
Comment (michel-laterman) [inline internal/pkg/config/timeouts.go:66]: @blakerouse @jsoriano added `checkin_max_poll` to timeouts. It only applies to client requested `poll_timeout` values (it does not effect `checkin_long_poll`).
I've extended the config unit test and smoke integration test to verify it works as expected.

Open to feedback for the default value, not sure if we want to use 2h or not

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module github.com/elastic/fleet-server/v7
 
-go 1.19
+go 1.20
```
Comment (jsoriano) [inline go.mod:3]: Should this change go on its own PR?

File: internal/pkg/config/timeouts.go (hunk 3 above)
Comment (jsoriano) [reply internal/pkg/config/timeouts.go:66]: Looks good to me, maybe we can start with a lower upper limit. Reducing it later may be breaking.

Should we add this new setting to the documentation?

**Review by jsoriano — APPROVED** [review]
Nice :+1:

File: internal/pkg/config/timeouts.go (hunk 3 above)
Comment (blakerouse) [reply internal/pkg/config/timeouts.go:66]: 2h seems high, I agree that lowering it would become breaking where raising it would not. So I would start at 1 hour, being its a setting it can be adjusted.

**Review by blakerouse — APPROVED** [review]
Looks good!

**Comment by jlind23** [comment]
@blakerouse do you think we can use the integration fwk you are working on in order to test this change? Two tests one where the checkin body contains the new field and one where it does not contain it would be great.
